utils/pkcs11helpers.go

```diff
@@ -0,0 +1,430 @@
+/*
+   Copyright The ocicrypt Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package utils
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/miekg/pkcs11"
+	"github.com/pkg/errors"
+	pkcs11uri "github.com/stefanberger/go-pkcs11uri"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	// OAEPSHA1Params describes the OAEP parameters with sha1 hash algorithm; needed by SoftHSM
+	OAEPSha1Params = &pkcs11.OAEPParams{
+		HashAlg:    pkcs11.CKM_SHA_1,
+		MGF:        pkcs11.CKG_MGF1_SHA1,
+		SourceType: pkcs11.CKZ_DATA_SPECIFIED,
+		SourceData: nil,
+	}
+	// OAEPSHA1Params describes the OAEP parameters with sha256 hash algorithm
+	OAEPSha256Params = &pkcs11.OAEPParams{
+		HashAlg:    pkcs11.CKM_SHA256,
+		MGF:        pkcs11.CKG_MGF1_SHA256,
+		SourceType: pkcs11.CKZ_DATA_SPECIFIED,
+		SourceData: nil,
+	}
+	// keep this order: try sha256 first, then fall back to sha1 if sha256 doesn't work
+	OAEPParams = []*pkcs11.OAEPParams{OAEPSha256Params, OAEPSha1Params}
+)
+
+// Pkcs11KeyFile describes the format of the pkcs11 (private) key file
+type Pkcs11KeyFile struct {
+	Pkcs11 struct {
+		Uri string `yaml:"uri"`
+	} `yaml:"pkcs11"`
+}
```

Comment:
I was trying to use YAML primarily because of readability but also because Kubernetes uses YAML files extensively. So far it's not much in this file, but it could grow when we start supporting HSM where possibly environment variables may have to be set.

Reply:
yea... my only worry is that if we were to expose need to expose this interface as GRPC, protobuf, they seem to be more friendly with JSON out of the box , like https://godoc.org/github.com/golang/protobuf/proto#Properties.

I have the impression that this struct may be part of a remote API call, but I may be wrong. 

Given this other perspective, what are your thoughts. I am okay with both, I don't feel strongly for either but just wanted to share the perspective.